src/rdkafka.h

```diff
@@ -1310,19 +1328,19 @@ rd_kafka_conf_res_t rd_kafka_conf_set(rd_kafka_conf_t *conf,
  * \p events is a bitmask of \c RD_KAFKA_EVENT_* of events to enable
  * for consumption by `rd_kafka_queue_poll()`.
  */
-RD_EXPORT
-void rd_kafka_conf_set_events(rd_kafka_conf_t *conf, int events);
```

Comment:
Existing APIs must not be changed since it breaks both the API and ABI guarantee.

The config scope validation is mainly useful for the generic _conf_set() APIs since they are typically exposed directly to the end-user.
The specific conf set functions, like this one and the ones for callbacks, are always added programmatically and it is sufficient to document what setting applies to which client type.

Reply:
Well the ABI is easily fixed by recompiling the code. I do think that anyone synch-ing with upstream code will all least recompile their stuff, especially if it's statically linked. Since C doesn't have a throw mechanism, unless you change the return type, it's impossible to validate. Also it's not just the config setters that are in question, it's also the callback setters which need this feature. This proposed change doens't really change the existing code, so I think it should be acceptable that at least a recompile should be done.

Reply:
ABI guarantee means that the application does **not** need to be recompiled when upgrading the library, this is how all shared libraries work on all the major platforms, Linux, osx, windows.
This is not up for discussion.

Regarding code-level setters, such as callback setters: yes, it would be good if they had a return value, but for historical reasons they don't and we can't change them.
My take on this is that it is not a critical concern since these are code-level setters, we can document our way around it to make sure the application developer will not use them for the wrong instance type.

Reply:
well in that case would you consider having a second set of API's which return an error? No ABI/API breakage.

Reply:
Another alternative is to leave the API as-is and have the return type be changed via a compile flag...actually this entire feature could then be compile controlled.

Reply:
Example:
```
# ifdef __HAS_RETURN__
# define RET_TYPE int
# define RETURN(val) return ((int)val) 
# else
# define RET_TYPE void
# define RETURN(val) return
# endif
```
Then you can use 
```
RET_TYPE some_function()
{
int a;
RETURN(a);
}
```

Reply:
That is not ABI or API safe, so we would need to provide two different libraries, and
we definitely don't want to go down that path.

Generally speaking, I see value in your proposed API changes in this PR, and if we were redesigning the API this would definitely go in, but we're not there, so each new API needs to be weighed careful, pros and cons, and if there is a viable workaround we typically reject API additions. Why? Because we're stuck with APIs "forever", and they thus need to be designed in a fashion to be future proof, which is hard. Add to that the API footprint size increasing and the maintenance cost, in KLOC, tests and support.

Specifically, if it stands between adding a second alternative function  for all conf setter functions, or adding a line to the documentation of each such function on what type of instance it should be used on, and the end-user is not directly affected, then we will choose the lesser approach of documenting it.
With documentation in place we can consider using set_dr_msg_cb() on a consumer config object a bug in the application, which it is.
Programmatic input does not need to be validated, only documented.
User input, on the other hand, needs proper validation and enforcement, and that's why I think the conf_set(name, value) interface should be changed for the new behaviour, but without affecting existing applications.

Adding a new conf_res_t enum will trigger a warning in any application that uses a switch-case on the conf_set() return value, we don't want that, so if we can solve that by using UNKNOWN (which a consumer property practically is on a producer, and vice versa) and providing a proper error string for the end-user, then we're fine.

Reply:
@edenhill what about my second proposal with the compile directives? (And I agree that having 2 sets of APIs is ugly.